google-cloud-logging/lib/google/cloud/logging/log/list.rb

```diff
@@ -0,0 +1,155 @@
+# Copyright 2017 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "delegate"
+
+module Google
+  module Cloud
+    module Logging
+      class Log
+        ##
+        # Log::List is a special case Array with additional values.
+        class List < DelegateClass(::Array)
+          ##
+          # If not empty, indicates that there are more records that match
+          # the request and this value should be passed to continue.
+          attr_accessor :token
+
+          ##
+          # @private Create a new Log::List with an array of log names.
+          def initialize arr = []
+            super arr
+          end
+
+          ##
+          # Whether there is a next page of logs.
+          #
+          # @return [Boolean]
+          #
+          # @example
+          #   require "google/cloud/logging"
+          #
+          #   logging = Google::Cloud::Logging.new
+          #
+          #   logs = logging.logs
+          #   if logs.next?
+          #     next_logs = logs.next
+          #   end
+          #
+          def next?
+            !token.nil?
+          end
+
+          ##
+          # Retrieve the next page of logs.
+          #
+          # @return [Sink::List]
+          #
+          # @example
+          #   require "google/cloud/logging"
+          #
+          #   logging = Google::Cloud::Logging.new
+          #
+          #   logs = logging.logs
+          #   if logs.next?
+          #     next_logs = logs.next
+          #   end
+          #
+          def next
+            return nil unless next?
+            ensure_service!
+            grpc = @service.list_logs token: token, resource: @resource,
+                                      max: @max
+            self.class.from_grpc grpc, @service
+          end
+
+          ##
+          # Retrieves all log names by repeatedly loading {#next} until
+          # {#next?} returns `false`. Calls the given block once for each log
+          # name, which is passed as the parameter.
+          #
+          # An Enumerator is returned if no block is given.
+          #
+          # This method may make several API calls until all log names are
```

Comment:
Maybe state explicitly how this compares with `each` (i.e. makes several API calls and loads all results, vs just iterating over the results in the current page)